v-next/hardhat/src/internal/cli/init/project-structure.ts

```diff
@@ -0,0 +1,55 @@
+import path from "node:path";
+
+import { HardhatError } from "@nomicfoundation/hardhat-errors";
+import { findUp, readUtf8File } from "@nomicfoundation/hardhat-utils/fs";
+import { findClosestPackageJson } from "@nomicfoundation/hardhat-utils/package";
+
+const JS_CONFIG_FILENAME = "hardhat.config.js";
+const TS_CONFIG_FILENAME = "hardhat.config.ts";
+const CJS_CONFIG_FILENAME = "hardhat.config.cjs";
+const CTS_CONFIG_FILENAME = "hardhat.config.cts";
+
+export async function isCwdInsideProject() {
+  return (
+    (await findUp(TS_CONFIG_FILENAME)) !== undefined ||
+    (await findUp(CTS_CONFIG_FILENAME)) !== undefined ||
+    (await findUp(CJS_CONFIG_FILENAME)) !== undefined ||
+    (await findUp(JS_CONFIG_FILENAME)) !== undefined
+  );
+}
+
+export async function getUserConfigPath() {
+  const tsConfigPath = await findUp(TS_CONFIG_FILENAME);
+  if (tsConfigPath !== null) {
+    return tsConfigPath;
+  }
+
+  const ctsConfigPath = await findUp(CTS_CONFIG_FILENAME);
+  if (ctsConfigPath !== null) {
+    return ctsConfigPath;
+  }
+
+  const cjsConfigPath = await findUp(CJS_CONFIG_FILENAME);
+  if (cjsConfigPath !== null) {
+    return cjsConfigPath;
+  }
+
+  const jsConfigPath = await findUp(JS_CONFIG_FILENAME);
+  if (jsConfigPath !== null) {
+    return jsConfigPath;
+  }
+
+  throw new HardhatError(HardhatError.ERRORS.GENERAL.NOT_INSIDE_PROJECT);
+}
```

Comment:
We have `findClosestHardhatConfig` that replaces both of these functions.

Reply:
But doesn't 'findClosestHardhatConfig' only search for the hardhat.config.ts file? Here we also have .js, .cjs and .cts files.
Also, 'findClosestHardhatConfig' is not currently exported from hh-utils. Should we export it and maybe change the logic to handle all the possible file extensions?

@schaable looping you in because of hh-utils. Wdyt?

Reply:
I wouldn't add it to hh-utils because the package doesn't really know about hardhat; all the utils are meant to be generic.

Reply:
@ChristopherDedominici I think that's actually a strong point for reusing it. If we recreate this logic here, it can easily be out of sync with the other implementation, which is the one used to load the config.

Regarding the actual extensions:

- `.js` is supported there.
- `.cjs` and `.cts` don't seem as important in v3, as it's designed for ESM projects. In v2 the config required CJS because it was imported by hardhat v2, which was cjs.

Reply:
As agreed offline on the approach. Implementation done in [this commit](https://github.com/NomicFoundation/hardhat/pull/5425/commits/63daf1f0286aaa0c25a8f3599deead2ee414020f)